type-tests/files/createSlice.typetest.ts

```diff
@@ -460,3 +463,89 @@ const value = actionCreators.anyKey
   expectType<ActionCreatorWithPayload<string>>(wrappedSlice.actions.success)
   expectType<ActionCreatorWithoutPayload<string>>(wrappedSlice.actions.magic)
 }
+
+{
+  interface TestState {
+    foo: string
+  }
+
+  interface TestArg {
+    test: string
+  }
+
+  interface TestReturned {
+    payload: string
+  }
+
+  interface TestReject {
+    cause: string
+  }
+
+  const slice = createSlice({
+    name: 'test',
+    initialState: {} as TestState,
+    reducers: withSpecial => ({
+      normalReducer(state, action: PayloadAction<string>) {
+        expectType<TestState>(state)
+        expectType<string>(action.payload)
+      },
+      testInfer: withSpecial.asyncThunk(
```

Comment:
Which name would we use for code & documentation for this builder object? `withSpecial` could be a start, but I'm not 100% sold yet.

Reply:
I'd like to throw in some options: `extra` _(has continuity with the `extra` property elsewhere in the lib, and the mental model matches)_, `enhancer`, and `addOn` _(my preferred)_.

Reply:
I believe we should not use `builder` here, as they work completely different.

The existing `builder` object just has methods that are called and the return type of the callback function accepting the builder doesn't matter.
In this case, it is quite the opposite: Just calling a method on this callback argument does nothing until you actually return it.

So, from a naming standpoint, it's the question how to read it fluently. Some of the current options:

```js
createSlice({
  name: test,
  reducers: builder => ({ someName: builder.asyncThunk( /*...*/ ) } ),
  reducers: extra => ({ someName: extra.asyncThunk( /*...*/ ) }) ,
  reducers: enhancer => ({ someName: enhancer.asyncThunk( /*...*/ ) }) ,
  reducers: addOn => ({ someName: addOn.asyncThunk( /*...*/ ) }) ,
  // adding a "with" prefix:
  reducers: withBuilder => ({ someName: withBuilder.asyncThunk( /*...*/ ) } ),
  reducers: withSpecial => ({ someName: withSpecial.asyncThunk( /*...*/ ) }) ,
  reducers: withEnhancer => ({ someName: withEnhancer.asyncThunk( /*...*/ ) }) ,
  reducers: withEnhanced => ({ someName: withEnhanced.asyncThunk( /*...*/ ) }) ,
  reducers: withAddOn => ({ someName: withAddOn.asyncThunk( /*...*/ ) }) ,
})
```

Personally, I'm not a fan of the `special` in `withSpecial`, but I like how the `reducers: withXXX =>` reads.
Currently, I quite like `withBuilder`, as it is *different* from the `builder`, but not *too* different. Also I think that `withEnhanced` sounds nice. 
I'm not a fan of `extra` though, as for me it's quite different from the thunk `extra` and might even lead to something like:

```js
createSlice({
  name: test,
  reducers: extra => ({ someName: extra.asyncThunk((arg, {extra}) {
    // shadowed variable alarm in here!
  }) }) ,
})
```

Reply:
I don't get the purpose of the `with` prefix, tbh.  This isn't a HOC or anything, and it _is_ a "builder"-type API.  It's just a _different_ builder than the one for `extraReducers`.

Reply:
Yeah, I totally forgot about the HOC-association of `with`. That's not too good then. Generally, I think it would be good to have something that distinguishes it from the existing builder api, because I believe that people will get that stuff mixed up.

Reply:
Agreed. Just not sure what a good option is atm.

Reply:
Hm. Just an imperative verb instead of a noun? `build` or `create`?

Reply:
Actually, that's half of my question: what _is_ the difference in builder methods here?

`extraReducers` has `builder.{addCase,addMatcher,addDefaultCase}`.  

What are the options here?  It looks like passing a callback here replaces the object notation, as with `extraReducers`.  How do we define a "normal" case reducer here?

Reply:
This notation gives you a set of tools (right now, just one), but expects you to return the same "reducers" object as you would have before, without the callback. 
A normal reducer in this context, would look just like before - see a few lines above this comment:

https://github.com/reduxjs/redux-toolkit/pull/637/files#diff-d3f6ba32c4f3b495b630095111a617e9R488

Using this way is necessary because while the existing builder notation is nice, it doesn't accumulate a final return value (and forcing that might result in more circular imports) - so if we used that, there would be no way from a TS perspective to know all the reducer/action names that were added, so we couldn't map over those and not create a valid type for `slice.action`

Reply:
I think that's why I personally prefer using 'addOn'. It makes sense as long as you can't also do `reducers: addOn => ({ thing: addOn.reducer() })`, and the addOn is just used for `thing: addOn.asyncThunk()`. That being said, if that doesn't mesh with everyone, I agree that `create` is probably the way to go and probably makes more sense than addOn either way.

Reply:
I've named it `create` for now.